Automattic/woocommerce-payments
# fix proposal for Apple pay issue w/ WordPress in subfolder

Fixes #3565 

#### Changes proposed in this Pull Request

- Parse URL from `get_site_url()`, stripping down any paths or subfolders, so the actual domain is passed down to the Apple Pay verification request.
- Better document things in the Apple Pay verification error notice.

#### Testing instructions

Test things breaking with the latest release of WooCommerce Payments, and then update the plugin at the end to see the fix in action:

1. Go to https://jurassic.ninja > More options > Launch Multisite on subdirs.
2. Wait for the creation of a multisite WP installation, then at the top bar, go to **My Sites > Network Admin > Plugins**, and network install these plugins from WP.org:
    - WooCommerce.
    - WooCommerce Payments.
    - WooCommerce Payments Dev Tools (download the zip from the repository and upload it to plugins).
3. Create a new site on a subfolder under **Sites > Add new**.
4. Go to the dashboard of the recently added site, and navigate to the WCPay Dev plugin.
5. Make sure all options are disabled, except for "Enable API request redirection".
6. Expose your server to a public URL using Ngrok or JT and update the "Redirect API requests to:" field to point to your public server URL. (Something like `https://{SERVER_URL}/wp-json/`).
7. Make sure your server is configured with live secrets from the `WCPay Dev (Express)` Stripe platform account, since the Apple domain verification does not work in test mode.
8. Since the verification doesn't work with a dev mode account, you'll need to connect to a live mode account, feel free to use mine (Ricardo's): [`acct_1ITcPcR03O9UZoSU`](https://dashboard.stripe.com/connect/accounts/acct_1ITcPcR03O9UZoSU).
    1. Click "Finish setup" from WooCommerce Payments onboarding, and finish the Jetpack authorization.
    2. Don't proceed further with KYC.
    3. From your local server's DB (`wcpay_accounts` table) or WP CLI, connect your blog ID (located in the dev tools plugin) with the `acct_1ITcPcR03O9UZoSU` Stripe account. The account record needs to be set in live mode (`is_live` = `1`).
    4. From the WCPay Dev plugin, clear the account cache to confirm you're successfully connected to `acct_1ITcPcR03O9UZoSU` live account.
9. As soon as you enable WooCommerce Payments, after you refresh the page, you should see the error notice below:
    - <img width="100%" alt="Apple Pay domain verification error notice" src="https://user-images.githubusercontent.com/5509901/171512778-6db72483-0745-43a9-8af3-ed7070e6bf9c.png">
    - This is because WCPay is passing the value from `get_site_url()`, which has a subfolder in it.
10. Now, with all the information that caused the issue, build the fix from this fork: 
     - `git clone https://github.com/ironicmoka/woocommerce-payments.git && cd woocommerce-payments`
     - `git checkout apple-pay-folder-wordpress-fix`
     - `npm install`
     - and `npm run build:release` to create a new zip for WCPay.
12. Upload the generated zip into the network plugins: **My Sites > Network Admin > Plugins**.
13. When prompted, replace the existing plugin with the new uploaded zip, and then activate WooCommerce Payments.
14. Trigger the domain registration again by disabling express checkouts, saving settings, and then enabling it again.
15. Error notice should be gone and domain should be verified (Apple Pay button should now show up on Safari).
